components/eam/src/physics/crm/cospsimulator_intr.F90

```diff
@@ -0,0 +1,3184 @@
+  ! SUBROUTINE setcosp2values
+  ! ######################################################################################
+#ifdef USE_COSP
+  subroutine setcosp2values(Nlr_in,use_vgrid_in,csat_vgrid_in,Ncolumns_in,cosp_nradsteps_in)
+    use mod_cosp,             only: cosp_init 
+    use mod_cosp_config,      only: vgrid_zl, vgrid_zu, vgrid_z
+    use mod_quickbeam_optics, only: hydro_class_init, quickbeam_optics_init
+    ! Inputs
+    integer, intent(in) :: Nlr_in             ! Number of vertical levels for CALIPSO and Cloudsat products
+    integer, intent(in) :: Ncolumns_in        ! Number of sub-columns
+    integer, intent(in) :: cosp_nradsteps_in  ! How often to call COSP?
+    logical, intent(in) :: use_vgrid_in       ! Logical switch to use interpolated, to Nlr_in, grid for CALIPSO and Cloudsat
+    logical, intent(in) :: csat_vgrid_in      !
+    
+    ! Local
+    logical :: ldouble=.false.
+    logical :: lsingle=.true. ! Default is to use single moment
+    integer :: i,k
+
+    prsmid_cosp  = pres_binCenters
+    prslim_cosp  = pres_binEdges
+    taumid_cosp  = tau_binCenters
+    taulim_cosp  = tau_binEdges
+    srmid_cosp   = calipso_binCenters
+    srlim_cosp   = calipso_binEdges
+    sza_cosp     = parasol_sza
+    dbzemid_cosp = cloudsat_binCenters
+    dbzelim_cosp = cloudsat_binEdges
+    htmisrmid_cosp = misr_histHgtCenters
+    htmisrlim_cosp = misr_histHgtEdges
+    taumid_cosp_modis = tau_binCenters
+    taulim_cosp_modis = tau_binEdges
+    reffICE_binCenters_cosp = reffICE_binCenters
+    reffICE_binEdges_cosp   = reffICE_binEdges
+    reffLIQ_binCenters_cosp = reffLIQ_binCenters
+    reffLIQ_binEdges_cosp   = reffLIQ_binEdges
+                                  
+    ! Initialize the distributional parameters for hydrometeors in radar simulator. In COSPv1.4, this was declared in
+    ! cosp_defs.f.
+    if (cloudsat_micro_scheme == 'MMF_v3.5_two_moment')  then
+       ldouble = .true. 
+       lsingle = .false.
+    end if
+    call hydro_class_init(lsingle,ldouble,sd)
+    call quickbeam_optics_init()
+
+    ! DS2017: The setting up of the vertical grid for regridding the CALIPSO and Cloudsat products is 
+    !         now donein cosp_init, but these fields are stored in cosp_config.F90.
+    !         Additionally all static fields used by the individual simulators are set up by calls
+    !         to _init functions in cosp_init.
+    ! DS2019: Add logicals, default=.false., for new Lidar simuldators (Earthcare (atlid) and ground-based
+    !         lidar at 532nm)
+    call COSP_INIT(Lisccp_sim, Lmodis_sim, Lmisr_sim, Lradar_sim, Llidar_sim, LgrLidar532, &
+         Latlid, Lparasol_sim, Lrttov_sim, radar_freq, k2, use_gas_abs, do_ray,              &
+         isccp_topheight, isccp_topheight_direction, surface_radar, rcfg_cloudsat,           &
+         use_vgrid_in, csat_vgrid_in, Nlr_in, pver, cloudsat_micro_scheme)
+
+    ! Set number of sub-columns, from namelist
+    nscol_cosp = Ncolumns_in
```

Comment:
Are these subcolumns the same as the subcolumn feature at the physics levels? or is this something internal to COSP?

Reply:
Internal to COSP.